pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -205,39 +201,48 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
+		configapi.SchemeGroupVersion,
+	)
+
+	if len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupExtensions)) > 0 {
+		resourceEncodingConfig.SetVersionEncoding(
+			configapi.APIGroupExtensions,
+			unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"},
+			extensions.SchemeGroupVersion,
+		)
 	}
 
-	// enable this if extensions API is enabled (or batch or autoscaling, since they persist to extensions/v1beta1 for now)
-	// TODO: replace this with a loop over configured storage versions
-	extensionsEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupExtensions)) > 0
-	batchEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupBatch)) > 0
-	autoscalingEnabled := len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupAutoscaling)) > 0
-	if extensionsEnabled || autoscalingEnabled || batchEnabled {
-		// TODO: replace this with a configured storage version for extensions once configuration exposes this
-		extensionsStorageVersion := unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"}
-		databaseStorage, err := NewEtcdStorage(etcdClient, extensionsStorageVersion, options.EtcdStorageConfig.KubernetesStoragePrefix)
-		if err != nil {
-			return nil, fmt.Errorf("Error setting up Kubernetes extensions server storage: %v", err)
-		}
-		storageDestinations.AddAPIGroup(configapi.APIGroupExtensions, databaseStorage)
-		storageVersions[configapi.APIGroupExtensions] = extensionsStorageVersion.String()
+	if len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupBatch)) > 0 {
+		resourceEncodingConfig.SetVersionEncoding(
+			configapi.APIGroupBatch,
+			unversioned.GroupVersion{Group: batch.GroupName, Version: "v1"},
+			batch.SchemeGroupVersion,
+		)
+	}
+
+	if len(configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupAutoscaling)) > 0 {
+		resourceEncodingConfig.SetVersionEncoding(
+			configapi.APIGroupAutoscaling,
+			unversioned.GroupVersion{Group: autoscaling.GroupName, Version: "v1"},
+			autoscaling.SchemeGroupVersion,
+		)
+	}
+
+	etcdConfig := storagebackend.Config{
+		Prefix:     options.EtcdStorageConfig.KubernetesStoragePrefix,
+		ServerList: options.EtcdClientInfo.URLs,
+		KeyFile:    options.EtcdClientInfo.ClientCert.KeyFile,
+		CertFile:   options.EtcdClientInfo.ClientCert.CertFile,
+		CAFile:     options.EtcdClientInfo.CA,
 	}
+	storageFactory := genericapiserver.NewDefaultStorageFactory(etcdConfig, "", kapi.Codecs, resourceEncodingConfig, master.DefaultAPIResourceConfigSource())
+	// the order here is important, it defines which version will be used for storage
+	storageFactory.AddCohabitatingResources(extensions.Resource("jobs"), batch.Resource("jobs"))
```

Comment:
@mfojtik @soltysh so we are going to store (or rather continue storing) jobs and hpa with the extensions api, is that right?